### DISCUSSION TITLE
Merge with blender-dactyl-keyboard

### DISCUSSION THREAD

**Discussion by Looooopy** [joshreve/dactyl-keyboard#57] [Ideas]
[I see: I was late in the game and @joshreve already found it on reddit and that he was named in the readme of that fork :) ]

I found a nice branch of tshort/dactyl-keyboard there he used blender to produce the output, the render speed was much slower (which is bad) but it generated a really smooth finish, I think it was worth waiting for :) 

See for your self [blender-dactyl-keyboard](https://github.com/SeanGilroy/blender-dactyl-keyboard) maybe you guys could work together to get things going?

**Comment by joshreve** [comment]
I'm already looking at this a little.  It isn't top of my list, since I have some more base functions I'm still fixing up.  The first step would be to create a blender helper like I have for cadquery and solidpython.  Once that was in place, it would really just be generation order, then an extra smooth step.  It looks like the smoothed blender geometry uses a lot of the same techniques, but stops to smooth before cutting square key holes and adding the key mount geometry.  I've reached out to the them and they have no problem with a similar technique being added to this repo, but unfortunately I have several other areas to clean up first. 

If anyone is interested in creating the blender helper that matches the interface of the other two, that would be a huge help.

**Reply by Looooopy** [reply]
Nice!

I can't jump in just yet I need to at least build my own first keyboard before so I get some kind of knowledge how everything fits together.
Good way to find both the gain and the pain :)

**Reply by joshreve** [reply]
Totally agree.  Focus on getting something functional first.  I am a fan of hot swap so you can upgrade more easily.  Without trying something it it is harder to decide what more complex features you want.

Definitely going to look at it, but I have to put functionality first so timeline is uncertain unless someone has time to jump in.

**Reply by BansheeHero** [reply]
Definitely support the author's goals. Python is perfect for this approach using helper files.
Having an extra engine could have impact for editing the solids later. (Like sculpting or combining with other objects.)
I actually like the low polygon look, but it does not fit 3d printing well as each polygon needs sanding individually. Perhaps something that other materials could resolve (PLA and PETG are not solvable using user-friendly chemicals).

In the same spirit I am expanding the `generate_configuration.py` in my repo:
- Refactor the save_config() from comments to DOCSSTRINGS
- Added a `--help` argument calling the content of DOCSTRING as help for the user..
- Working on adding help to each geometry section that will enable `--interactive` mode where use can select values instead of adjusting JSONs.
- Add differential engine that would allow `--partial` mode.
    - Second JSON will only contain the changed values. 
    - This would load the partial config instead and use the `generate_config.py` to fill-out the rest.
    - This will effectively allow people to maintain different already established configs. (Full JSON for best compatibility, partial JSON that would gather parameters maintained by the main repository.)

Also README.md is a bit bloated now so I am trying to refactor it. All in good faith and maybe worth a pull later.

**Reply by joshreve** [reply]
You may want to look at the refactor branch.  It is new, but already works when executed with code.  I haven't fully setup JSON support or cleaned out all the deprecated files yet.  I'm planning to move to it long term, but have a PCB project to finish before returning my priority to this repo.  The big change is breaking the design into pieces for easier modification.  It is also introducing json capable dataclasses.  I am hoping to use the dataclass interface to autogenerate a web parameter UI.  I've done it in the past with Flask for a RPi based robot, but haven't looked at it in a while.

Either way, definitely interested in what you come up with.

**Reply by BansheeHero** [reply]
Thank you for the info, I will check it out. Good luck on your other efforts.
